src/binaryServerStarter.ts

```diff
@@ -0,0 +1,150 @@
+            + 'Run anyways?', 'Yes', 'No')
+            .then((val: string) => {
+              if (val === 'Yes') {
+                resolve(binaryExecutable);
+              }
+              reject("The binary XML language server is not trusted");
+            });
+        });
+      }
+    });
+}
+
+/**
+ * Returns the path to the LemMinX binary
+ *
+ * Downloads it if it is missing
+ *
+ * @returns The path to the LemMinX binary
+ * @throws If the LemMinX binary can't be located or downloaded
+ */
+async function getServerBinaryPath(): Promise<string> {
+  const config: WorkspaceConfiguration = getXMLConfiguration();
+  let binaryPath: string = config.get("server.binary.path");
+  if (binaryPath) {
+    if (fs.existsSync(binaryPath)) {
+      return Promise.resolve(binaryPath);
+    }
+    window.showErrorMessage('The specified XML language server binary could not be found. Using the default binary...');
+  }
+  let server_home: string = path.resolve(__dirname, '../server');
+  let binaries: Array<string> = glob.sync(`**/lemminx-${os.platform()}*`, { cwd: server_home });
+  const JAR_AND_HASH_REJECTOR: RegExp = /(\.jar)|(hash)$/;
+  binaries = binaries.filter((path) => { return !JAR_AND_HASH_REJECTOR.test(path) });
+  if (binaries.length) {
+    return new Promise<string>((resolve, reject) => {
+      resolve(path.resolve(server_home, binaries[0]));
+    });
+  }
+  // Download it, then return the downloaded binary's location
+  return downloadBinary();
+}
+
+/**
+ * Downloads LemMinX binary under the `server` directory and returns the path to the binary as a Promise
+ *
+ * @returns The path to the LemMinX binary
+ * @throws If the LemMinX binary download fails
+ */
+async function downloadBinary(): Promise<string> {
+  window.setStatusBarMessage('Downloading XML language server binary...', 2000);
+  return new Promise((resolve, reject) => {
+    const handleResponse = (response: http.IncomingMessage) => {
+      const statusCode = response.statusCode;
+      if (statusCode === 303) {
+        http.get(response.headers.location, handleResponse);
+      } else if (statusCode === 200) {
+        // This is probably the problem in Theia TODO:
+        const serverBinaryPath = path.resolve(__dirname, '../server', getServerBinaryName());
+        const serverBinaryFileStream = fs.createWriteStream(serverBinaryPath);
+        response.pipe(serverBinaryFileStream);
+        serverBinaryFileStream.on('finish', () => {
+          serverBinaryFileStream.on('close', () => {
+            fs.chmodSync(serverBinaryPath, "766");
+            resolve(serverBinaryPath);
+          });
+          serverBinaryFileStream.close();
+        });
+      } else {
+        reject('Server binary download failed');
+      }
+    }
+    http.get(`http://localhost:8080/lemminx-redirect/${os.platform()}`, handleResponse)
+      .on('error', () => {
+        reject('Server binary download failed');
```

Comment:
There's no details whatsoever about why the download failed. Not helpful

Reply:
Thanks for catching this, should have error messages now.